ff/serving/job/materialize_job.go

```diff
@@ -0,0 +1,148 @@
+package job
+
+import (
+	"fmt"
+	"sync"
+)
+
+type Runner interface {
+	Run() (CompletionStatus, error)
+}
+
+type MaterializedChunkRunner struct {
+	Materialized MaterializedFeatures
+	Table        OnlineTable
+	ChunkSize    int
+	ChunkIdx     int
+}
+
+type CompletionStatus interface {
+	Complete() bool
+	String() string
+	Wait() error
+	Err() error
+}
+
+type MaterializedFeatures interface {
+	NumRows() (int, error)
+	IterateSegment(begin int, end int) (FeatureIterator, error)
+}
+
+type OnlineTable interface {
+	Set(entity string, value interface{}) error
+	Get(entity string) (interface{}, error)
+}
+
+type FeatureIterator interface {
+	Next() bool
+	Err() error
+	Entity() string
+	Value() interface{}
+}
+
+func (m *MaterializedChunkRunner) Run() (CompletionStatus, error) {
+	done := make(chan interface{})
+	resultSync := &ResultSync{}
```

Comment:
You can actually just make the CopyCompletionWatcher up here and make a private method SetDone(err) that also closes the channel